design/ceph/node-loss-rbd-rwo.md

```diff
@@ -0,0 +1,105 @@
+
+## Proposed Solution
+
+> Note: This solution requires minimum kubernetes version 1.26.0
+
+The kubernetes feature [Non-Graceful Node Shutdown](https://kubernetes.io/blog/2022/12/16/kubernetes-1-26-non-graceful-node-shutdown-beta/) is available starting in K8s 1.26 to help improve the volume recovery during node loss. When a node is lost, the admin is required to add the taint `out-of-service` manually to the node. After the node is tainted, K8s will:
+
+- Remove the volume attachment from the lost node
+- Delete the old pod on the lost node
+- Create a new pod on the new node
+- Allow the volume to be attached to the new node
+
+Once this taint is applied manually, rook will create a [NetworkFence CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md). The [csi-addons operator](https://github.com/csi-addons/kubernetes-csi-addons) will then blocklist the node to prevent any ceph rbd client on the lost node from writing any more data.
+
+After the new pod is running on the new node and the old node which was lost comes back, rook will delete the [NetworkFencr CR](https://github.com/csi-addons/kubernetes-csi-addons/blob/main/docs/networkfence.md).
+
+example of taint to be applied to lost node:
+
+```console
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoExecute
+# or
+kubectl taint nodes <node-name> node.kubernetes.io/out-of-service=nodeshutdown:NoSchedule
+```
+
+Example of a NetworkFence CR that the Rook operator would create when a node is lost:
+
+```yaml
+apiVersion: csiaddons.openshift.io/v1alpha1
+kind: NetworkFence
+metadata:
+  name: <name> # We will keep the name the same as the node name
+  namespace: <ceph-cluster-namespace>
+spec:
+  driver: <driver-name> # for us it will `rook-ceph.rbd.csi.ceph.com`
+  fenceState: <fence-state> # For us it will be `Fenced`
+  cidrs:
+    - <nodeIP>
+  secret:
+    name: <csi-rbd-provisioner-secret-name>
+    namespace: <ceph-cluster-namespace>
+  parameters:
+    clusterID: <ceph-cluster-namespace>
+```
+
+> Note: This will be enabled by default in rook, in the case for some reason user wants to disable this feature in rook can edit `rook-ceph-operator-config` configmap and update the `ROOK_WATCH_FOR_NODE_FAILURE: "false"`.
+
+## How get which NodeIp to block
+
+There are multiple networking options available for example, host Networking, Pod networking, Multus etc. This make it difficult to know which NodeIp address to blockList.
+For this we'll follow the following approach which will work for all networking options(but still will not work for external cluster which nodeIps):
+
+1. Get the `volumesInUse` from the node which has the taint `out-of-service`
```

Comment:
How do we query the volumesInUse?

Reply:
To blocklist the ip we need the image, to get the image from the pv we'll compare it with pv.

Reply:
There is no field called `volumesInUse`, right? Maybe just put it in separate words then.
```suggestion
1. Get the volumes in use from the node which has the taint `out-of-service`
```

Reply:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#nodestatus-v1-core

Actually there is a field but I am not sure what it refers too or how it is constructed. 
We'll need to make sure that it is the same across k8s release, has a consistent pattern and stable for us to use it.

```
 volumesAttached:
  - devicePath: ""
    name: kubernetes.io/csi/rook-ceph.rbd.csi.ceph.com^0001-0009-rook-ceph-0000000000000002-0ea0490d-2b28-40f8-a812-5e98de2b3a55
  volumesInUse:
  - kubernetes.io/csi/rook-ceph.rbd.csi.ceph.com^0001-0009-rook-ceph-0000000000000002-0ea0490d-2b28-40f8-a812-5e98de2b3a55
 ```
 
 @subhamkrai have you looked deeper into this ? 
 
 cc @Madhu-1

Reply:
yes there is the field, if a volume is mounted on that node

Reply:
Perfect, I had overlooked the `volumesInUse`